lib/utils/pagination/pagination.go

```diff
@@ -0,0 +1,50 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pagination
+
+import "github.com/gravitational/trace"
+
+// IteratePages will iterate through pages of a paging function and call the given function to each element of the results.
+// The map function can return false to short circuit and return from the function early.
+func Iterate[T any](initialToken string, pageFn func(string) ([]T, string, error), mapFn func(T) (bool, error)) error {
+	pageToken := initialToken
+	for {
+		var resources []T
+		var err error
+		resources, pageToken, err = pageFn(pageToken)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		for _, resource := range resources {
+			ok, err := mapFn(resource)
```

Comment:
I'm not sure I quite understand this. The way I see this, the purpose of `pageFn` is to produce a list of resources given a `pageToken` and the purpose of `mapFn` is to run some arbitrary unit of work against each resource produced by `pageFn`. Without something like `mapFn` in place I still have a large amount of boilerplate to add into `pageFn` which doesn't really simplify things enough for me to use a utility like this.

It's stuff like [this](https://github.com/gravitational/teleport/blob/master/lib/services/local/saml_idp_service_provider.go#L194) I'm trying to make less repetitive. Every time I need to do something like iterate through these service providers (for example) and do something like collect service provider names or check for uniqueness, I have this pattern each time:

```go
var pageToken string
for {
  var resources []someResource
  var err error
  resources, pageToken, err = <service-call>
  if err != nil {
    return trace.Wrap(err)
  }

  for _, resource := range resources {
    <perform-some-work>
  }

  if pageToken == "" {
    break
  }
}
```

In order to do the same thing without having something like `mapFn`, I feel like `pageFn` would basically just encapsulate the entirety of the `for` block above, which is mostly what I'm trying to eliminate. I could potentially just aggregate the results into one array and then iterate on that result later (which is what I believe you suggested before), but that seems somewhat wasteful w/r/t things like the uniqueness check example above where I don't really need to store anything, just check it and throw it away.

A better name for this function might be `ForEach` instead of `Iterate`. Does this make sense, or am I missing something here?

Reply:
That linked example, using this pagination function as I have it written, would look like:

```go
// Make sure no other service provider has the same entity ID.
pagination.Iterate("", s.ListSAMLIdPServiceProviders, func() (bool, error) {
  if listSp.GetName() != sp.GetName() && listSp.GetEntityID() == sp.GetEntityID() {
    return true, trace.BadParameter("%s %q has the same entity ID %q", types.KindSAMLIdPServiceProvider, listSp.GetName(), listSp.GetEntityID())
  }
})
```

To me, that doesn't seem overly complex, What do you think?

Reply:
I think that's harder to understand than writing out a bit more, and I still think `mapFn` is redundant. The `(bool, error)` return is not obvious either, I'd have to dig into Iterate to figure it out.

Version without `mapFn`, for comparison:

```go
// `ctx`, `pageSize` and `sp` captured by the closure.
err := pagination.Iterate("" /* initialPageToken */, func(pageToken string) ([]types.SAMLIdPServiceProvider, string, error) {
  providers, nextPageToken, err := s.ListSAMLIdPServiceProviders(ctx, pageSize, pageToken)
  if err != nil { // Note: We could go straight to looping if we wanted.
    return nil, "", trace.Wrap(err)
  }

  for _, p := range providers {
    if p.GetName() != sp.GetName() && p.GetEntityID() == sp.GetEntityID() {
      return nil, "", trace.BadParameter("%s %q has the same entity ID %q", types.KindSAMLIdPServiceProvider, p.GetName(), p.GetEntityID())
    }
  }

  return providers, nextPageToken, nil
})
// handle err
```

Sure, it has more lines, but it's straightforward to read and reason about. This kind of code I'm OK repeating.

I think some of the convenience you get from `mapFn` is only there because your starting point are hand-wrapped RPCs. Once you try to apply this to a "pure" RPC (for example, a direct call to [ListDevices](https://github.com/gravitational/teleport/blob/a82b267c87a2cbd2d7df8cb4f5da1be9fa528ac1/api/proto/teleport/devicetrust/v1/devicetrust_service.proto#L79)), `mapFn`'s convenience goes out of the window, because the paginated function has to be wrapped in a `func` as well.

ListDevices example (aka non-wrapped RPC):

```go
err := pagination.Iterate("", func(pageToken string) ([]*devicepb.Device, string, error) {
  req.PageToken = pageToken
  resp, err := devicesClient.ListDevices(ctx, req) // RPC signature doesn't match `pageFn`.
  if err != nil {
    return nil, "", trace.Wrap(err)
  }
  return resp.Devices, resp.NextPageToken, nil
})
// handle err
```

Non-generic version for comparison, just because:

```go
var allProviders []types.SAMLIdPServiceProvider
err := pagination.Iterate("", func(pageToken string) (string, error) {
  providers, nextPageToken, err := s.ListSAMLIdPServiceProviders(ctx, pageSize, pageToken)
  
  // Filter as desired, use a hand-written for loop, etc.
  for _, provider := range providers {
    if /* something */ {
      return nil, trace.BadParameter("...")
    }
    allProviders = append(allProviders, provider)
  }

  return nextPageToken, trace.Wrap(err)
})
// handle err
```

I don't want to hold you on this any longer, so feel free to ping me directly if you want to chat a bit more.